src/state.rs

```diff
@@ -591,6 +443,183 @@ impl DemandProxyState {
     }
 }
 
+/// A Dns Resolver is responsible for the DNS resolving task for given hostnames
+#[derive(Default, Debug, Clone)]
+struct DnsResolver {
+    // Map of resolved hostnames.
+    resolved: Arc<RwLock<HashMap<String, ResolvedDns>>>,
+
+    // Map of in-progress resolution requests.
+    in_progress: Arc<Mutex<HashMap<String, Arc<Notify>>>>,
```

Comment:
I'm thinking that: is it possible that putting both `resolved` and `in_progress` in a struct such as `ResolvedDnsStore`? Then we just need to define `resolved_dns: Arc<RwLock<ResolvedDnsStore>>` as a field in `DnsResolver`.

Reply:
I kept them separate initially, because of how they're used. The `resolved` map is the fast-path and shouldn't need a write lock for the majority cases (once some amount of resolution has been done). Accessing `in_progress` needs a write lock every time it's accessed, because it's a "get or create" operation.  That said, you could benchmark it with one lock vs 2 and see if it makes much of a difference in practice.